internal/parser/parser.go

```diff
@@ -207,25 +207,30 @@ func findPort(svc *corev1.Service, wantPort kongstate.PortDef) (*corev1.ServiceP
 
 func getUpstreams(
 	log logrus.FieldLogger, s store.Storer, serviceMap map[string]kongstate.Service) []kongstate.Upstream {
-	var upstreams []kongstate.Upstream
+	upstreamDedup := make(map[string]struct{})
+	var empty struct{}
+	upstreams := make([]kongstate.Upstream, 0, len(serviceMap))
 	for _, service := range serviceMap {
-		var targets []kongstate.Target
-		port, err := findPort(&service.K8sService, service.Backend.Port)
-		if err == nil {
-			targets = getServiceEndpoints(log, s, service.K8sService, port)
-		} else {
-			log.WithField("service_name", *service.Name).Warnf("skipping service - getServiceEndpoints failed: %v", err)
-		}
+		name := fmt.Sprintf("%s.%s.%s.svc", service.Backend.Name, service.Namespace, service.Backend.Port.CanonicalString())
+		if _, exists := upstreamDedup[name]; !exists {
```

Comment:
Just so we're clear, as we decided to go with this implementation as opposed to one of the more performant (but significantly longer time investment) options as described in https://github.com/Kong/kubernetes-ingress-controller/pull/1702#discussion_r689674761 we are ultimately adding a fair amount of compute overhead by added this lookup on every item.

Before we consider this comment resolved and merge this PR, I would just like to suggest we take some cycles to run this code with an incredibly large set of services (10k?) in the before and after state so that we have some kind of loose idea of what that performance difference actually may be.

In the short term (prior to any major refactor) if we find we need to get more performance out of this area, we may be able to use something like a [bloom filter](https://en.wikipedia.org/wiki/Bloom_filter) to improve the lookup performance in the base case where there is no duplicate.

Reply:
https://github.com/Kong/kubernetes-ingress-controller/issues/1709